### PR TITLE
[MXNET-1164] Generate the document for cpp-package using Doxygen

### PR DESCRIPTION
**Pull request by leleamol**
## Description ##
The PR includes changes to Doxyfile for generating documents for cpp-package.  Also modified the index.md file in api/c++ directory to include brief build instructions and links to the correct classes in mxnet::cpp namespace

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [y] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [y] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)
Verified that the doxygen generates the docs for cpp-package.

### CODE REVIEW

**Comment by ankkhedia** [comment]
@leleamol Thanks for your contribution!

@mxnet-label-bot [pr-awaiting-review, C++]

**Comment by ankkhedia** [comment]
@nswamy Could you please take a look into this PR?

**Comment by kalyc** [comment]
ping @leleamol ! could you please provide the generated doxygen files?

**Comment by leleamol** [comment]
@nswamy There will be a doxygen file for each hpp and h file in the cpp-package folder.

Should I paste all those files here?

**Comment by stu1130** [comment]
@nswamy ping

**Comment by leleamol** [comment]
@aaronmarkham

File: docs/Doxyfile [hunk 1]
```diff
@@ -770,7 +770,7 @@ WARN_LOGFILE           =
 # spaces.
 # Note: If this tag is empty the current directory is searched.
 
-INPUT                  = include src/common
+INPUT                  = include src/common cpp-package/include/mxnet-cpp
```
Comment (aaronmarkham) [inline docs/Doxyfile]: Is this combining the C and C++ API together in one Doxygen output?

Comment (leleamol) [reply docs/Doxyfile]: I dont have the dev web server that is publicly accessible for preview.

**Comment by Roshrini** [comment]
@leleamol Can you retrigger the CI for this PR? 
@mxnet-label-bot Update [pr-awaiting-testing]

**Comment by aaronmarkham** [comment]
@leleamol Please add `3rdparty` to the `EXCLUDE` directive in `Doxyfile`. That should help retrigger CI as well as provide the additional benefit of limiting the doxygen run to relevant files. 

I don't know why it runs on 3rdparty - you'd think that the INPUT directive limits things, but if you look at the logs you see:
```
Generating file sources...
Generating code for file 3rdparty/dlpack/include/dlpack/dlpack.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/any.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/array_view.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/base.h...
Generating code for file 3rdparty/mshadow/mshadow/base.h...
Generating code for file include/mxnet/base.h...
Generating code for file 3rdparty/tvm/nnvm/include/nnvm/base.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/blockingconcurrentqueue.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/common.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/concurrency.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/concurrentqueue.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/config.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/data.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/endian.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/input_split_shuffle.h...
Generating code for file 3rdparty/dmlc-core/include/dmlc/io.h...
...
```

I tried excluding the 3rdparty folder and that does seem to work... it does a pass through the folder making an exclusion list.

WDYT?

**Comment by leleamol** [comment]
Hi @aaronmarkham I have updated the Doxyfile and excluded the 3rdparty folder.
I verified that cpp-package doc gets correctly generated and 3rdparty folders are also getting excluded.

**Comment by leleamol** [comment]
@mxnet-label-bot remove [ pr-awaiting-testing ]

**Pull request merged**